workspace-manager-clienttests/src/main/java/scripts/testscripts/ControlledBigQueryDatasetLifecycle.java

```diff
@@ -0,0 +1,181 @@
+      throws Exception {
+    super.doSetup(testUsers, workspaceApi);
+    // Note the 0th user is the owner of the workspace, pulled out in the super class.
+    assertThat(
+        "There must be at least three test users defined for this test.",
+        testUsers != null && testUsers.size() > 2);
+    this.writer = testUsers.get(1);
+    this.reader = testUsers.get(2);
+  }
+
+  @Override
+  protected void doUserJourney(TestUserSpecification testUser, WorkspaceApi workspaceApi)
+      throws Exception {
+
+    ControlledGcpResourceApi ownerResourceApi =
+        ClientTestUtils.getControlledGcpResourceClient(testUser, server);
+
+    // Add a writer and reader to this workspace
+    logger.info(
+        "Adding {} as writer and {} as reader to workspace {}",
+        writer.userEmail,
+        reader.userEmail,
+        getWorkspaceId());
+    workspaceApi.grantRole(
+        new GrantRoleRequestBody().memberEmail(writer.userEmail), getWorkspaceId(), IamRole.WRITER);
+    workspaceApi.grantRole(
+        new GrantRoleRequestBody().memberEmail(reader.userEmail), getWorkspaceId(), IamRole.READER);
+
+    logger.info("Waiting 15s for permissions to propagate");
+    TimeUnit.SECONDS.sleep(15);
+
+    // Create a cloud context
+    String projectId = CloudContextMaker.createGcpCloudContext(getWorkspaceId(), workspaceApi);
+    logger.info("Created project {}", projectId);
+
+    // Create a shared BigQuery dataset
+    CreatedControlledGcpBigQueryDataset createdDataset = createDataset(ownerResourceApi);
+    UUID resourceId = createdDataset.getResourceId();
+
+    // Retrieve the dataset resource
+    logger.info("Retrieving dataset resource id {}", resourceId.toString());
+    GcpBigQueryDatasetResource fetchedResource =
+        ownerResourceApi.getBigQueryDataset(getWorkspaceId(), resourceId);
+    assertEquals(createdDataset.getBigQueryDataset(), fetchedResource);
+
+    BigQuery ownerBqClient = ClientTestUtils.getGcpBigQueryClient(testUser, projectId);
+    BigQuery writerBqClient = ClientTestUtils.getGcpBigQueryClient(writer, projectId);
+    BigQuery readerBqClient = ClientTestUtils.getGcpBigQueryClient(reader, projectId);
+
+    // Workspace owner can create a table in this dataset
+    Table table = createTable(ownerBqClient, projectId);
+    String tableName = table.getTableId().getTable();
+
+    // Workspace reader can read the table
+    var readTable = readerBqClient.getTable(table.getTableId());
+    assertEquals(table, readTable);
+    logger.info("Read table {} as workspace reader", tableName);
+
+    // Workspace reader cannot modify tables
+    Table readerUpdatedTable = table.toBuilder().setDescription("A new table description").build();
+    assertThrows(
+        BigQueryException.class,
+        () -> readerBqClient.update(readerUpdatedTable),
+        "Workspace reader was able to modify a table");
+    logger.info("Workspace reader could not modify table {} as expected", tableName);
+
```

Comment:
How complicated would it be to use queries to test access in addition to touching the table metadata?
It would be a nice-to-have

Reply:
Good call!